src/LuaVector2.cpp

```diff
@@ -0,0 +1,256 @@
+		} else if (!strcmp(attr, "y")) {
+			v->y = luaL_checknumber(L, 3);
+		} else {
+			luaL_error(L, "Index '%s' is not available: use 'x' or 'y'", attr);
+		}
+
+	} else {
+		luaL_error(L, "Expected vector2, but type is '%i'", lua_type(L,2));
+	}
+	LuaVector2::PushToLua(L, *v);
+	return 1;
+}
+
+static int l_vector_index(lua_State *L)
+{
+	const vector2d *v = LuaVector2::CheckFromLua(L, 1);
+	if (lua_type(L, 2) == LUA_TSTRING) {
+		const char *attr = luaL_checkstring(L, 2);
+		if (!strcmp(attr, "x")) {
+			lua_pushnumber(L, v->x);
+			return 1;
+		} else if (!strcmp(attr, "y")) {
+			lua_pushnumber(L, v->y);
+			return 1;
+		}
+	}
+	lua_getmetatable(L, 1);
+	lua_pushvalue(L, 2);
+	lua_rawget(L, -2);
+	lua_remove(L, -2);
+	return 1;
+}
+
+static int l_vector_normalised(lua_State *L)
+{
+	const vector2d *v = LuaVector2::CheckFromLua(L, 1);
+	LuaVector2::PushToLua(L, v->NormalizedSafe());
+	return 1;
+}
+
+static int l_vector_length(lua_State *L)
+{
+	const vector2d *v = LuaVector2::CheckFromLua(L, 1);
+	lua_pushnumber(L, v->Length());
+	return 1;
+}
+
+static int l_vector_magnitude(lua_State *L)
+{
+	const vector2d *v = LuaVector2::CheckFromLua(L, 1);
+	lua_pushnumber(L, v->LengthSqr());
+	return 1;
+}
+
+static int l_vector_angle(lua_State *L)
+{
+	const vector2d *v = LuaVector2::CheckFromLua(L, 1);
+	lua_pushnumber(L, atan2(v->x, v->y));
+	return 1;
+}
+
+static int l_vector_rotate(lua_State *L)
+{
+	vector2d *v = LuaVector2::CheckFromLua(L, 1);
+	const double angle = luaL_checknumber(L, 2);
+	LuaVector2::PushToLua(L, v->Rotate(angle));
+	return 1;
+}
+
+static int l_vector_rot_90_left(lua_State *L)
+{
+	const vector2d *v = LuaVector2::CheckFromLua(L, 1);
+	const vector2d rot(-v->y, v->x);
+	LuaVector2::PushToLua(L, rot);
+	return 1;
+}
+
+static int l_vector_rot_90_right(lua_State *L)
+{
+	const vector2d *v = LuaVector2::CheckFromLua(L, 1);
+	const vector2d rot(v->y, -v->x);
+	LuaVector2::PushToLua(L, rot);
+	return 1;
+}
+
+static luaL_Reg l_vector_lib[] = {
+	{ "new", &l_vector_new },
+	{ "unit", &l_vector_unit },
+	{ "length", &l_vector_length },
+	{ 0, 0 }
+};
+
+static luaL_Reg l_vector_meta[] = {
+	{ "__tostring", &l_vector_tostring },
+	{ "__add", &l_vector_add },
+	{ "__sub", &l_vector_sub },
+	{ "__mul", &l_vector_mul },
+	{ "__div", &l_vector_div },
+	{ "__unm", &l_vector_unm },
+	{ "__index", &l_vector_index },
+	{ "__newindex", &l_vector_new_index },
+	{ "normalised", &l_vector_normalised },
+	{ "normalized", &l_vector_normalised },
+	{ "unit", &l_vector_unit },
+	{ "length", &l_vector_length },
+	{ "magnitude", &l_vector_magnitude},
+	{ "dot", &l_vector_magnitude},
+	{ "rotate", &l_vector_rotate },
+	{ "angle", &l_vector_angle },
+	{ "left", &l_vector_rot_90_left },
+	{ "right", &l_vector_rot_90_right },
+	{ 0, 0 }
+};
+
+const char LuaVector2::LibName[] = "vector2";
+const char LuaVector2::TypeName[] = "vector2";
+
+void LuaVector2::Register(lua_State *L)
+{
+	LUA_DEBUG_START(L);
+
+	luaL_newlib(L, l_vector_lib);
```

Comment:
To make `Vector2()` create a new table, you'll need this syntax after `luaL_newlib(...)`:
```c++
lua_newtable(L);
lua_pushcfunction(L, &l_vector_new);
lua_setfield(L, -2, "__call");
lua_setmetatable(L, -2);
```
You'll also need to rewrite your function to support passing the `Vector2` lib table as the first argument, as the syntax of the metamethod is `function __call(tab, ...)`.